[chip] New component

mbrookes opened the pull request:
- [x] PR has ~~tests~~ / docs demo, and is linted.
- [x] Commit and PR titles begin with [ComponentName], and are in imperative form: "[Component] Fix leaky abstraction".
- [x] Description explains the issue / use-case resolved, and auto-closes the related issue(s) (http://tr.im/vFqem).

New `Chip` component.

Features:
- Text only and "Contact" avatars.
- Contact Avatars can be image, icon (svg or font) or text.
- Optional delete icon.
- Focus shadow, click shading.
- Touch and keyboard events supported.

![screen cap chips](https://cloud.githubusercontent.com/assets/357702/14315614/5ae4f6c2-fbf6-11e5-8ecb-f5417cdd86f0.gif)

Still to do:
- ~~Agree the API - currently supports both props and composability for `label` and `Avatar`. As a new component we might want to only support composability.~~
- ~~Decide on required event callbacks. I took a kitchen sink approach, but we might want to narrow it down (or add others, such as drag events).~~
- ~~`PropTypeDescription` is refusing to work, and I'm not sure why, so Ive removed it from the docs `Page.js` for now.~~
- ~~Refactoring based on @callemall/material-ui review :sweat_smile:~~
- ~~Tests.~~ (Thanks @nathanmarks!)
- ~~Delete button not working in Firefox. (zIndex?)~~ (Thanks @newoga!)

Closes #1119.

mbrookes committed:
[Chip] Add new component

mbrookes committed:
[EnhancedButton] Reset button margin and padding, Esc key removes focus

Default button margin and padding varies by browser, so reset it to zero.

Can still be overidden by style prop.

nathanmarks commented:
Looks pretty!

nathanmarks commented on src/internal/EnhancedButton.js:
```diff
@@ -273,6 +273,8 @@ class EnhancedButton extends React.Component {
       cursor: disabled ? 'default' : 'pointer',
       textDecoration: 'none',
+      margin: 0,
+      padding: 0,
```
Can we remove the reset in other components using `EnhancedButton` ?

mbrookes replied:
Should be able to - I'll probably do it in another PR.

oliviertassinari replied:
I'm wondering about the side effect of this change. I will try each page :grin:.

newsiberian commented:
Great work! small question: any chances have editable `chip`s. Like for an "add new tag" purpose?

mbrookes commented:
What is an editable chip?

newsiberian commented:
> What is an editable chip?

Sorry, I've checked specification and there is nothing about it. Could it be possible to pass a `children` node to chip to override default `label` field, like we have now in other components, like button for example?

newoga commented:
Looks good @mbrookes! I'll review in more detail soon!

oliviertassinari commented on docs/src/app/components/pages/components/Chip/ExampleSimple.js:
```diff
@@ -0,0 +1,96 @@
+import {blue300, indigo900} from 'material-ui/styles/colors';
+
+const style = {
+  margin: '4px',
```
You can simply use `4`.

oliviertassinari commented on docs/src/app/components/pages/components/Chip/README.md:
```diff
@@ -0,0 +1,5 @@
+## Chip
+
+[Chips](https://www.google.com/design/spec/components/chips.html) represent complex entities in small blocks, such as a contact.
```
Would be better to break this line.

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,338 @@
+    },
+    chip: {
+      backgroundColor: state.clicked ? pressedColor : backgroundColor,
+      borderRadius: '16px',
```
`16` would do it.

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,338 @@
+      paddingRight: '12',
+      whiteSpace: 'nowrap',
+    },
+    root: {
```
I'm not sure we need this.

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,338 @@
+}
+
+class Chip extends React.Component {
+  static muiName = 'Chip';
```
What is the use case of this property?

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,338 @@
+  };
+
+  static defaultProps = {
+    onBlur: () => {},
```
I'm wondering what the best choice between using a default function and checking that the function property exist.

nathanmarks replied:
@oliviertassinari `typeof xxx` is faster than calling a `noop()` IIRC, but both are so fast it's inconsequential, 100s of millions of ops per second.

What style do you prefer?

mbrookes replied:
> I'm wondering what the best choice between using a default function and checking that the function property exist.

I had one, then both, then the other.

This approach also gives us an explicit default in the property docs, although I'm fine with that being blank if explicit checking is preferred. 

This is also make for cleaner code in the event handlers compared to checking existence of each prop.

oliviertassinari replied:
> This approach also gives us an explicit default in the property docs

I'm not sure this is a real advantage.

> This is also make for cleaner code in the event handlers compared to checking existence of each prop.

Yeah, I feel like it's the real advantage.

Looks like React Bootstrap and React Native are following the *checking* path:
- https://facebook.github.io/react-native/docs/pickerios.html#content
- https://github.com/facebook/react-native/blob/master/Libraries/Picker/PickerIOS.ios.js#L76

They are probably doing it for performances reason. I would personally do the same, and use the *checking existence* path. 
@nathanmarks That's also saving some function allocations. That's a microoptimisation :grin:.

nathanmarks replied:
@oliviertassinari hahahaha good point. Though this one is even more inconsequential than the `new Something()` vs `Object.create()` performance debate :smile: Unless the events are firing millions of times per second (imagine that application!) 

Let's go with the microoptimisation! :grinning:

mbrookes replied:
Since the benefit of checking in inconsequential, I'll leave it as is for the sake of simpler code.

oliviertassinari replied:
What about the other components?
I have the feeling that we are more often using the *checking* way. Should we clean the one that are not using this pattern?

mbrookes replied:
I've seen both, although you''re probably right about checking be more common, and it would be good to be consistent, but this doesn't add a new inconsistency.

At the moment, many (but not all) of the checks are single line `if` statements:

```js
if (this.props.someEventProp) this.props.someEventProp(event)
```
If we accept that single line `if`s are not good style, then the alternative is:
```js
if (this.props.someEventProp) {
  this.props.someEventProp(event)
}
```
which starts to look very verbose compared setting defaults, and just using:
```
this.props.someEventProp(event)
```
especially when you have a large number of events.

So my proposal is that we:

* Use this style here
* Convert others as we find them, (and possibly do a clean-up across the board at some point as low-priority)
* Implement an eslint rule for no single line `ifs`

But I'm open to alternatives (and lets not get bogged down on this - Chip can be changed to match some other code style in the future as easily as any other component. :smile: )

oliviertassinari replied:
I completly agree with having a rule for no single line `if`s.
Regarding how verbose our check is, I don't mind. I would rather chose the performance path.

> but this doesn't add a new inconsistency

True, let's keep it like this and move forward. We can easily fix the inconsistency later.

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,338 @@
+    const {prepareStyles} = this.context.muiTheme;
+    const styles = getStyles(this.props, this.context, this.state);
+
+    const chipStyle = prepareStyles(Object.assign(styles.chip, this.props.chipStyle));
```
Why not destructuring the `chipStyle`?

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,338 @@
+        onKeyboardFocus={this.handleKeyboardFocus}
+        style={rootStyle}
+      >
+        <div {...other} style={chipStyle}>
```
I think that we should use `{...other}` on the root component: `EnhancedButton `.

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,338 @@
+      null;
+
+    // If the`avatar` prop isn't set, check if the first child is an Avatar
+    if (!avatar) {
```
I'm wondering if we should support both redondant option:
```jsx
<Chip
  label="Composed Image Avatar"
  avatar={<Avatar icon={<Avatar src="images/kolage-128.jpg" />} />}
/>
```

```jsx
<Chip>
  <Avatar src="images/kolage-128.jpg" />
  Composed Image Avatar
</Chip>
```

oliviertassinari replied:
I think that we should only support one of the two options. I feel like the second option is better. It's providing more possibility without much overhead.

mbrookes commented:
Q: Do we need the `deletable` prop, or should I simply test for the presence of `onRequestClose`?

oliviertassinari commented:
@mbrookes Is the following use case valid ?
- `deletable={false}`
- the `Chip` is focused
- the user is using the delete key

That trigger `onRequestClose`.

mbrookes commented:
@oliviertassinari - I'm not sure it is - if the chip needs to be deletable, it presumably needs to support touch devices?

oliviertassinari commented:
@mbrookes In that case, I'm fine with just using the `onRequestClose`. Do what you think is the best :+1:.

mbrookes commented:
Okay, I'll try that see how it looks. I'll update the prop description of that's how it works best.

mbrookes commented:
@oliviertassinari - I implement that and updated the prop docs.

I also:

* Fixed the Firefox delete, and one other issues thanks to @newoga, 
* Simplified the code some more
* Allow Esc key removing focus added to EnhancedButton
* Move the callback prop docs to Enhanced button, and imported from there
* Added basic tests

oliviertassinari commented on docs/src/app/components/pages/components/Chip/Page.js:
```diff
@@ -0,0 +1,33 @@
+import chipReadmeText from './README';
+import ChipExampleSimple from './ExampleSimple';
+import chipExampleSimpleCode from '!raw!./ExampleSimple';
+import chipCode from '!raw!material-ui/lib/Chip/Chip';
```
Do we need the `lib` path here?

mbrookes replied:
Hm, no, you're right. We should probably change that in all the other components

oliviertassinari replied:
Sounds good to me.

oliviertassinari commented on docs/src/app/components/pages/components/Chip/Page.js:
```diff
@@ -0,0 +1,33 @@
+import chipCode from '!raw!material-ui/lib/Chip/Chip';
+import enhancedButtonCode from '!raw!material-ui/lib/internal/EnhancedButton';
+
+const description = 'Examples of Chips using an image, [Font Icon](/#/components/font-icon), ' +
```
We are enforcing ES6 template. Shouldn't eslint complain here?

mbrookes replied:
About what, sorry? :speak_no_evil:

nathanmarks replied:
Maybe there's an exception for multi line string concatenation?

mbrookes replied:
Looking at the docs for that rule, it does only gives examples of single-line concatenation, but I just tested, and that passes too. Odd!

oliviertassinari replied:
Odd indeed. As you point out. We follow this pattern everywhere else in the docs. I think we can forget about it in this PR.

mbrookes commented:
One more Q: Should the Chip include default left and right margin, so that it lays out nicely without additional styling? It seems like something you would need in any pretty-much any application of `Chip`.

nathanmarks commented:
Couple quick things:

1. Why `onRequestClose` if the prop is `deletable`? `onDelete` seems more appropriate.
2. The `delete` dialog is popping up twice for me in the docs example -- this is because both `onClick` and `onTouchTap` are set probably.

mbrookes commented:
>Why onRequestClose if the prop is deletable? onDelete seems more appropriate.

So the 'Request` bit of `onRequestClose` differentiates from `onClose` in that `Close` hasn't happened, it has only been requested, and it's up to the event handler to take the appropriate action. an alternative might be `onClickClose` or `onTouchTapClose`, as it's more descriptive of the event that has occurred.

As for `Close` vs `Delete` - I struggled with that, and wanted to discuss it, so glad you brought it up. `onXXXDelete` is semantically more correct, as a `Chip` doesn't 'close', but `onRequestClose` is standard for other components, so at least it's consistent. 

Happy to go with the consensus on both these point, including changing the other components to match if it comes to that.

> The delete dialog is popping up twice for me in the docs example -- this is because both onClick and onTouchTap are set on DeleteIcon probably.

How the heck did that happen?! That was the other @mbrookes!

nathanmarks commented:
@mbrookes tomato tomato I suppose, but I think `onRequestDelete` is more suitable. Consistency between components is irrelevant if the functionality represents something different. You close a drawer, you delete a chip.

mbrookes commented:
> How the heck did that happen?! That was the other @mbrookes!

Okay, I wasn't going mad! :smile: : 
https://github.com/mbrookes/material-ui/blame/chip/src/Chip/Chip.js#L231

nathanmarks commented:
@mbrookes HAHAHA.

Oh man, yeah -- LOL. Oops :dancers:

nathanmarks commented:
@mbrookes it sucks having to **require** users to use that plugin

nathanmarks commented:
@mbrookes it's because I added a test to make sure `onClick` works, but in the tests we're obviously using simulated events since this isn't selenium/webdriver so it doesn't trigger both. Doh.

oliviertassinari commented:
@mbrookes What is missing for this PR to be merged ?

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,271 @@
+      <DeleteIcon
+        color={styles.closeIcon.color}
+        style={styles.closeIcon}
+        onClick={this.handleCloseIconTouchTap}
```
Why adding onClick? I think that we don't have any example with this property.

nathanmarks replied:
@oliviertassinari It's getting removed, that was my mistake -- It's also causing a bug at the moment in the docs 👍

mbrookes replied:
Thanks @nathanmarks. @oliviertassinari - It's already gone in my local repo.

mbrookes commented:
> @mbrookes What is missing for this PR to be merged ?

@oliviertassinari I just need to tweak the styles slightly - that's what kicked off the whole `colorManipulator.js` rewrite. Something I was trying to do wasn't working as expected, and I fell down the rabbit hole when I discovered what a mess that module was. 

Now that's sorted, I can finish this up.

mbrookes commented:
One Q still outstanding - should I include any default right margin? Seems like every use case will need it, but not all of our components have it.

oliviertassinari commented:
> should I include any default right margin?

@mbrookes As far as I know, none of our components have a default margin. I would say no. I think that it should be the responsibility of the container to add it.

mbrookes commented:
Ok, thanks. I'll leave that as is.

jakeboone02 commented:
This component will frequently be part of a group of Chips, though, so if there is no default margin, I think you should at least provide some guidance about spacing between Chips on the docs.  The spec appears to use 12px between Chips, on the right and on the bottom for when they wrap to multiple rows.

On Apr 17, 2016, at 11:55, Matt Brookes <notifications@github.com> wrote:

Ok, thanks. I'll leave that as is.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub

mbrookes commented:
@jakeboone02 - `8px` between chips. For simplicity for the demo I put `4px` all round.

jakeboone02 commented:
@mbrookes From what I can tell, the spec isn't explicit about it (did I miss it?).  But [this example](https://www.google.com/design/spec/components/chips.html#chips-specs) from the Single-line chips / Non-deletable chips section has 12px between them:

![image](https://cloud.githubusercontent.com/assets/366438/14622653/72dc9936-0580-11e6-9996-64bf03ad585b.png)

mbrookes commented:
@jakeboone02 No, it isn't, but 8px is the defined spacing between buttons, and the spacing used in the examples if I"m not mistaken. In any case, you're free to use whatever spacing you wish.

jakeboone02 commented:
> you're free to use whatever spacing you wish

I agree!  ;)

Thanks.

nathanmarks commented:
You will use the spacing that I deem appropriate. Nothing more, nothing less.

nathanmarks commented:
So that'd be whatever you want.

nathanmarks commented:
@mbrookes squishy squashy

mbrookes commented:
Squashed all but the most recent 2 commits (while keeping updates to separate components in their own commit).

nathanmarks commented:
@mbrookes Safari 7 came out in 2013 and received mostly security updates after that. I think we can probably take that off the "Still to do" list.

nathanmarks commented:
Small thing -- do you think this should say `noop` instead of the empty anonymous function? https://dl.dropboxusercontent.com/s/68dx1hk2juv704d/2016-04-28%20at%207.41%20PM.png

nathanmarks commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,270 @@
+  };
+
+  static defaultProps = {
+    onBlur: () => {},
```
@mbrookes in another PR @oliviertassinari ended up going with `typeof` checks instead of creating a noop function. Should we have a consistent implementation throughout the lib? I'll let you guys hash it out 😄

alitaheri replied:
:+1: check before invocation. what if the user does this: `onBlue={shouldExplodeInBlur && () => explode()}`
now it explosed anyway :/

mbrookes commented:
> Safari 7 came out in 2013 and received mostly security updates after that. I think we can probably take that off the "Still to do" list.

Fixed. I was hamstrung at 7 due to corporate build, but wanted to ensure the issues with 7 were not present in current versions. I take it you've tested?

> do you think this should say noop instead of the empty anonymous function?

They're common across the lib. Not mine to argue.

> Should we have a consistent implementation throughout the lib?

It's already inconsistent - we hashed this out elsewhere. If someone wants to change it post-merge, feel free,

nathanmarks commented:
@mbrookes 

**note:** This looks like a problem with our `EnhancedButton` because I noticed that hitting `enter` with the `Choose an image` button example is completely broken, so we have an issue with keyboard accessibility for our `EnhancedButton` in general.

Basically -- I can't use enter to trigger a chip touchtap handler more than once without refocusing the button. (not correct behaviour, test here: https://jsfiddle.net/gakgbLb5/).

Out of scope of this PR, but something we should fix at some point.

nathanmarks commented:
Are we holding off on the merge until `0.15.0` has been released? I see this is marked `0.15.1`

mbrookes commented:
@nathanmarks Not sure why it's milestoned for 0.15.1 specifically, but who ever says so is responsible for all future rebases. 😆

nathanmarks commented:
Only bugs are out of scope of this PR, so looks good to me 👍

jakeboone02 commented:
In testing this, I found that `props.onRequestDelete` gets fired on right-click as well as left-click.  I have a multi-select component built on `react-virtualized-select`, using `Chip` as the `valueComponent`, and that behavior is a little jarring.  When I right-click I expect a context menu, not an action to be performed.  The `Button` components behave in the expected way, for example.

I could inspect the touchtap event in my `onRequestDelete` handler and determine what to do from there, but I wonder if a user would _ever_ expect a context menu-type click to remove an element from a list.  If they wouldn't, then handling the left-vs-right click decision inside this component would make it easier on developers like me. :)

mbrookes commented:
Thanks for the feedback Jake, I'll take a look. I'm not sure how we're handling this in other components.

nathanmarks commented:
@mbrookes I didn't catch that in my review -- but I agree with @jakeboone02 that we should get a context menu.

Until we have this setup with a multi-select/autocomplete in mui, maybe being less opinionated in that aspect is the best route. For example -- if chips link to something, I can't open them in a new tab using the context menu if the right click is hijacked.

jakeboone02 commented:
To clarify, the problem is only with the `DeleteIcon` component within the Chip.  Right-clicking on any other part of the Chip brings up a context menu as expected.

mbrookes commented:
@jakeboone02 I've tried in Chrome, Firefox and Safari, and in all cases I get the browser context menu when I right-click on the delete Icon.

jakeboone02 commented:
Oh, you know what...this might be an issue with `react-select` and not the `Chip` component.  I'll test some more.

jakeboone02 commented:
OK, looks like the issue is with `react-select` and not with the `Chip` component.  Sorry for the noise!

mbrookes commented:
@jakeboone02 - thanks for circling back and confirming. 👍

jakeboone02 commented:
Ugh...I apologize again.  The problem seems to be with both components.  The issue with `Chip` is that the `onRequestDelete` handler fires on _any_ mouse/touch event.  I think the reason you saw the right-click menu in your testing is that `Chip` doesn't do `event.preventDefault()`.

On the new "Chips" docs page, right-clicking a `DeleteIcon` brings up the alert saying "You clicked the delete button," _then_ brings up a context menu.

In my case, `onRequestDelete` is bound to a function that removes the chip from a multi-select value list.  So the chip is removed first, and then I get a context menu.

Again, I could inspect and react to the event in my own component, but `Chip` could be modified to prevent this upstream.  This is probably not exactly the right way to do it, but it works and shows the intent:

```diff
   handleTouchTapCloseIcon = (event) => {
+    // Only capture events from the primary mouse button or touch tap
+    if (event.nativeEvent.type === 'touchend' ||
+        (event.nativeEvent.type === 'mouseup' &&
+          event.nativeEvent.button === 0)) {
       // Stop the event from bubbling up to the `Chip`
       event.stopPropagation();
       this.props.onRequestDelete(event);
+    }
   };

```

mbrookes commented:
@jakeboone02 When I tested, I only got the context menu, not the alert for all three browsers that I tested, but perhaps it's OS dependant? (I'm on OS X).

Being specific about which button we're listening for seems reasonable though.

nathanmarks commented:
@mbrookes 

I can reproduce this on windows, I checked the event object too... `onTouchTap` is being fired from a bonafide right click event.

Looks like a bug in [`react-tap-event-plugin`](https://github.com/zilverline/react-tap-event-plugin). When I change all of the examples to `onClick` instead of `onTouchTap` in `ExampleSimple.js`, the issue is no longer present.

_Sigh_

nathanmarks commented:
@callemall/material-ui 

We have a bug in `react-tap-event-plugin` from what I can see.

I'm able to trigger this alert by right clicking in Windows 10 (tested in IE edge and chrome):

```
<button
  onTouchTap={() => alert('hello')}
>
  Hello
</button>
```

mbrookes commented:
@nathanmarks - I don't have a windows machine to test with. If this is 'PR: needs-revision' to apply a workaround, please could you do the honours?

nathanmarks commented:
@mbrookes Unfortunately there is no work-around for this -- technically this does not need a revision. The bug appears to be with `react-tap-event-plugin` and it is pretty severe.

nathanmarks commented:
@mbrookes I'll create a separate issue for it.

nathanmarks commented:
@mbrookes this is pretty tragic: https://github.com/callemall/material-ui/issues/4149

oliviertassinari commented on docs/src/app/components/pages/components/Chip/ExampleArray.js:
```diff
@@ -0,0 +1,67 @@
+import React from 'react';
```
`import React, {Component} from 'react';`?

newoga replied:
I noticed this too but saw that all the examples haven't been migrated yet. Maybe we could do them all in a separate PR? Doesn't matter to me though

nathanmarks commented:
@mbrookes Something I just realised:

```js
  handleMouseEnter = (event) => {
    this.handleFocus();
    this.props.onMouseEnter(event);
  };
```

This fires the `handleFocus()` function, which then fires the `props.onFocus` handler. The problem here is that we are telling the user that the element has been focussed, when in fact it has just been hovered. 

We should try to keep the two event handlers separate -- and use another function if they need to share behaviour.

oliviertassinari commented on docs/src/app/components/pages/components/Chip/ExampleArray.js:
```diff
@@ -0,0 +1,67 @@
+ */
+export default class ChipExampleSimple extends React.Component {
+
+  constructor(props) {
```
Couldn't we remove the `constructor` method and use a class property for the state?

nathanmarks replied:
People are requesting that we make examples compatible with no babel stages enabled. What do you think?

oliviertassinari replied:
Well, it depends on how well stage 1 / 2 / 3 are used. Let me have a look at the npm download stats.

oliviertassinari replied:
Here here the adoption rate compared to [es2015](https://www.npmjs.com/package/babel-preset-es2015):
 - [babel-preset-stage-3](https://www.npmjs.com/package/babel-preset-stage-3): 43%
 - [babel-preset-stage-2](https://www.npmjs.com/package/babel-preset-stage-2): 42%
 - [babel-preset-stage-1](https://www.npmjs.com/package/babel-preset-stage-1): 32%
 - [babel-preset-stage-0](https://www.npmjs.com/package/babel-preset-stage-0): 26%

mbrookes replied:
All our non-stateless examples are done this way...

oliviertassinari commented on docs/src/app/components/pages/components/Chip/ExampleArray.js:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  handleRequestDelete = (key) => {
+    if (key === 3) {
```
`3` feel like a magical number. Could we add a comment?

oliviertassinari commented on docs/src/app/components/pages/components/Chip/ExampleArray.js:
```diff
@@ -0,0 +1,67 @@
+/**
+ * An example of rendering multiple Chips from an array of values.
+ */
+export default class ChipExampleSimple extends React.Component {
```
Doesn't match the name of the file.

newoga commented:
Just making a note here based on discussion in gitter:

Aside from whether or not hover should trigger shadow by default, the only suggestion I have is to change the font size and weight to `14px` and `normal`/`400`.

mbrookes commented:
Fixed all the things. :)

mbrookes commented:
Just playing with this again, and something still doesn't feel right with the "no `onTouchTap` event" chips.

Should deletable chips have hover state, even though they have no click handler, i.e. have everything except shadow on click?) I have no clue! 😄 

Current behaviour is that they don't change appearance (or pointer) on hover, but they can by focused with the mouse or keyboard, and change color to the focused color.

Other implementations:
http://materializecss.com/chips.html
No states at all.

https://material.angularjs.org/latest/demo/chips
No hover state.
Changes color on focus.
No shadow when clicked.
No pointer change.

http://react-toolbox.com/#/components/chip
Only the delete icon has hover state and pointer.

So I think Chip behaves better than all of those...

nathanmarks commented:
@mbrookes wanna squash the 3 chip commits into 1?

mbrookes commented:
Done.

nathanmarks commented:
@callemall/material-ui can we get another set of eyes to review this (properly)? Then we can merge IMO.

oliviertassinari commented on docs/src/app/components/pages/components/Chip/ExampleArray.js:
```diff
@@ -0,0 +1,64 @@
+  },
+};
+
+let chipData = [
```
I would just inline this object in the `constructor`. We are not reusing it.

oliviertassinari commented on src/internal/EnhancedButton.js:
```diff
@@ -37,31 +37,139 @@ function listenForTabPresses() {
+     *
+     * @param {object} event `touch` event targeting the element.
+     */
+    onTouchEnd: PropTypes.func,
```
I'm wondering if we shouldn't `@ignore` the native callbacks from the docs:
- `onMouseDown`
- `onMouseEnter`
- `onMouseLeave`
- `onMouseUp`
- `onTouchEnd`
- `onTouchStart`

Still, I would keep `onTouchTap` as it's not a native event.

mbrookes replied:
Yes, this was from before I suggested we remove the native event callbacks from the component docs. If all that's left is `onTouchTap`, it might as well go in the `Chip` prop docs rather than separately imported from `EnhancedButton`.

oliviertassinari commented on docs/src/app/components/pages/components/Chip/ExampleSimple.js:
```diff
@@ -0,0 +1,105 @@
+          onTouchTap={handleTouchTap}
+          style={styles.chip}
+        >
+          <Avatar src={'images/ok-128.jpg'} />
```
What about?
```jsx
<Avatar src="images/ok-128.jpg" />
```

mbrookes replied:
😊

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,283 @@
+    this.props.onBlur(event);
+  };
+
+  handleFocus= (event) => {
```
`handleFocus = `

mbrookes replied:
🔎 👍 

(Surprised linting didn't catch that...)

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,283 @@
+        containerElement="div" // Firefox doesn't support nested buttons
+        disableTouchRipple={true}
+        disableFocusRipple={true}
+        ref="button"
```
What the ref is for?

mbrookes replied:
No longer needed.

oliviertassinari commented:
@mbrookes I have only see minor issues. That's a great PR 👍.

mbrookes commented:
@oliviertassinari Thanks!

Let me know if that looks okay, and I'll squash.

zachguo commented:
Is this going to work with Autocomplete, like rendering Chips in text field at least?

mbrookes commented:
@zachguo This is just the chip itself, but an input component would be a natural place to use it.

keremciu commented:
@mbrookes @zachguo is there any plan to build an example using autocomplete with chips? I think it is really important for chip component, I will try to build multiple selectable autocomplete with chips maybe we can talk on that?

nathanmarks commented:
@keremciu yes, just not immediately

nathanmarks commented:
@mbrookes squash the fix all the things

mbrookes commented:
@keremciu You could open an issue for that with relevant screenshots, links to other implementations etc.

mbrookes commented:
@nathanmarks Done.

mbrookes commented:
Any idea why Chip doesn't show in the default coverage report? (Also do we need a couple more tests?)

nathanmarks commented:
@mbrookes https://coveralls.io/builds/6118507/source?filename=src%2FChip%2FChip.js

mbrookes commented:
@nathanmarks Yeah, I found that - that's why I asked about adding tests (for those last couple of functions).

The question was why Chip doesn't show in the _default_ coverage report. https://coveralls.io/builds/6118507 Just seems odd not to include new modules, tested or otherwise.

nathanmarks commented:
@mbrookes must be a little bug -- will report it

oliviertassinari commented:
@mbrookes That looks us some time. That worth it 🎉 .

Petergatsby commented:
This is great- well needed! Thanks @tim

muxica commented:
@mbrookes @nathanmarks, in reference to @keremciu's comment on May 10th:

> is there any plan to build an example using autocomplete with chips? I think it is really important for chip component, I will try to build multiple selectable autocomplete with chips maybe we can talk on that?

Is there any update on when examples with Autocomplete and Chips will be made? This use case is of particular interest to myself and I presume many others as well.

mbrookes commented:
@muxica I don't believe anyone is working on it.

elodszopos commented:
+1 on the Autocomplete chips integration.

mbrookes commented:
For anyone still looking that hasn't spotted it in related projects: https://github.com/TeamWertarbyte/material-ui-chip-input
